KAFKA-5516: Formatting verifiable producer/consumer output in a similar fashion

ppatierno opened the pull request.

ppatierno committed:
Modified JSON order attributes in a more readable fashion

ppatierno committed:
Fixed documentation about old request.required.acks producer parameter
Modified JSON order attributes in a more readable fashion
Refactoring on verifiable producer to be like the verifiable consumer

asfgit commented:
Refer to this link for build results (access rights to CI server needed): 
https://builds.apache.org/job/kafka-pr-jdk8-scala2.12/5689/
Test FAILed (JDK 8 and Scala 2.12).

asfgit commented:
Refer to this link for build results (access rights to CI server needed): 
https://builds.apache.org/job/kafka-pr-jdk7-scala2.11/5703/
Test FAILed (JDK 7 and Scala 2.11).

ppatierno committed:
Removed unused imports

asfgit commented:
Refer to this link for build results (access rights to CI server needed): 
https://builds.apache.org/job/kafka-pr-jdk7-scala2.11/5704/
Test PASSed (JDK 7 and Scala 2.11).

asfgit commented:
Refer to this link for build results (access rights to CI server needed): 
https://builds.apache.org/job/kafka-pr-jdk8-scala2.12/5690/
Test PASSed (JDK 8 and Scala 2.12).

ppatierno committed:
Refactoring on the verifiable producer to be quite similar to the consumer implementation

asfgit commented:
Refer to this link for build results (access rights to CI server needed): 
https://builds.apache.org/job/kafka-pr-jdk7-scala2.11/5712/
Test PASSed (JDK 7 and Scala 2.11).

asfgit commented:
Refer to this link for build results (access rights to CI server needed): 
https://builds.apache.org/job/kafka-pr-jdk8-scala2.12/5698/
Test PASSed (JDK 8 and Scala 2.12).

ewencp commented on tools/src/main/java/org/apache/kafka/tools/VerifiableProducer.java:
```diff
@@ -242,66 +207,174 @@ public String getValue(long val) {
-        Map<String, Object> data = new HashMap<>();
-        data.put("name", "shutdown_complete");
-        return toJsonString(data);
+    @JsonPropertyOrder({ "timestamp", "name" })
```
I like the `JsonPropertyOrder` solution when it's being applied on top of an existing class. But overall this patch makes things more verbose (we've added almost 100 lines). That's not necessarily a problem, but would it maybe be simpler to just change the existing maps to be `LinkedHashMap`s so they preserve their order and then order the `put` calls the way we want them?

ppatierno replied:
Is this way more explicit than leveraging on the order of the `put` execution ?
Maybe there are a lot of lines of code because other than doing this I have updated the verifiable producer to be quite similar to the verifiable consumer. See how I changed the `createFromArgs()` method and adding `ProducerEvent` class with all derived classes for defining event to log (like already happens with the verifiable consumer). Before these changes this tools seemed to be written by two completely different persons (maybe it's true ?) without considering that they do something similar (even if in the opposite direction). :-)

ewencp replied:
Scanning through the diff my impression was that most of the additions were due to all the new POJOs. They don't really add value over just creating a map since we don't ever really use the objects other than for serialization. I think sticking to POJOs makes sense when you then have other code that takes advantage of the objects, gets proper type checking, etc. It just seems verbose here to me.

`createFromArgs` changes seem reasonable, I just wasn't sure about about the verbosity of the `ProducerEvent` approach.

ppatierno commented:
@ewencp as I said before I introduced the POJO in the VerifiableProducer in order to be more consistent with the VerifiableConsumer. In the current version they are completely different : while VerifiableConsumer has POJOs (see ConsumerEvent), VerifiableProducer doesn't. So the idea to add ProducerEvent and at same time fixing the output to be similar. In my understanding for what you are saying we should remove the POJOs from the VerifiableConsumer as well. Is that right ? Having these two applications much more similar is better than having two completely different implementations. Wdyt ?

ewencp commented:
@ppatierno Heh, those classes look like they were written by two different people because they were written by two different people :) I didn't review the VerifiableConsumer code, otherwise I might have noted the relative verbosity of the POJO approach when pretty much every class is used only once.

I'm fine with getting things consistent, with either solution. The increase in LOC is just very noticeable when you see it in patch/PR form.

ppatierno commented:
@ewencp in general it's better to follow one direction so having more consistent code. It's the reason why I decided to use the POJO for the VerifiableProducer. So is this PR acceptable ? :)

ewencp commented on tools/src/main/java/org/apache/kafka/tools/VerifiableProducer.java:
```diff
@@ -154,79 +153,45 @@ private static ArgumentParser argParser() {
+        Properties producerProps = new Properties();
+        if (configFile != null) {
+            try {
+                producerProps.putAll(Utils.loadProps(configFile));
```
Changing to `Utils.loadProps` won't be safe for the reason noted in the comment that was deleted when the duplicated `loadProps` method was removed. Since we still have upgrade tests that run against 0.8 brokers, we still need the duplicated code.

ewencp commented on tools/src/main/java/org/apache/kafka/tools/VerifiableProducer.java:
```diff
@@ -319,59 +392,69 @@ public void onCompletion(RecordMetadata recordMetadata, Exception e) {
+
+            producer.run(throttler);
+        } catch (ArgumentParserException e) {
+            parser.handleError(e);
```
It seems like we've changed the behavior where no arguments just produces the normal help screen? Or is `handleError` already doing that?

ppatierno replied:
At the top of the `main` there is the `parser.printHelp();` which print the normal help screen when no arguments are provided.

ewencp reviewed (commented):
@ppatierno Yeah, mostly looks good. I had a few more comments and once they're resolved we'll want to run this through the system tests before committing since the tests are the primary user of this code.

ppatierno committed:
Re-added removed loadProps method for compatibility and upgrade tests against 0.8.X.X

ewencp commented:
Hmm, I tried running some tests and `kafkatest.tests.core.transactions_test.TransactionsTest.test_transactions.failure_mode=clean_bounce.bounce_target=brokers` seems to be consistently hanging. Will need to look into why this change would have affected that test.

ppatierno committed:
Fixed order of producer properties initialization

ppatierno commented:
@ewencp I found and fixed the problem ! Enabling the debug on the test, I got the verifiable producer stucked so I tried to launch the console application using the same parameters as the test (i.e. topic, properties, ...). The problem was the "retries" which is 50 in the test (using a properties file) and it was overridden by 0 because I revert the order of applying properties from config file with properties from command line. With "retries" equals 0 I got this exception : 

`
Exception in thread "main" org.apache.kafka.common.KafkaException: Failed to construct kafka producer
	at org.apache.kafka.clients.producer.KafkaProducer.<init>(KafkaProducer.java:416)
	at org.apache.kafka.clients.producer.KafkaProducer.<init>(KafkaProducer.java:301)
	at org.apache.kafka.tools.VerifiableProducer.createFromArgs(VerifiableProducer.java:204)
	at org.apache.kafka.tools.VerifiableProducer.main(VerifiableProducer.java:455)
Caused by: org.apache.kafka.common.config.ConfigException: Must set retries to non-zero when using the idempotent producer.
	at org.apache.kafka.clients.producer.KafkaProducer.configureRetries(KafkaProducer.java:467)
	at org.apache.kafka.clients.producer.KafkaProducer.<init>(KafkaProducer.java:359)
`

Now I have fixed that, recovering the right order. 
The only concern is that there were no evidence of this exception during the test.

asfgit commented:
Refer to this link for build results (access rights to CI server needed): 
https://builds.apache.org/job/kafka-pr-jdk7-scala2.11/6351/
Test PASSed (JDK 7 and Scala 2.11).

asfgit commented:
Refer to this link for build results (access rights to CI server needed): 
https://builds.apache.org/job/kafka-pr-jdk8-scala2.12/6336/
Test PASSed (JDK 8 and Scala 2.12).

ijuma commented:
Did you mean to commit the debug logs?

ppatierno commented:
@ijuma what do you mean sorry ?

ijuma commented:
@ppatierno take a look at the pull request diff, it adds 11k lines due to some debug files that were added.

ppatierno commented:
@ijuma sorry my fault when I did git add and commit. I'm going to delete them immediately ! Sorry !

ppatierno committed:
Deleted debug logs file committed by mistake

asfgit commented:
Refer to this link for build results (access rights to CI server needed): 
https://builds.apache.org/job/kafka-pr-jdk7-scala2.11/6567/
Test PASSed (JDK 7 and Scala 2.11).

asfgit commented:
Refer to this link for build results (access rights to CI server needed): 
https://builds.apache.org/job/kafka-pr-jdk8-scala2.12/6552/
Test PASSed (JDK 8 and Scala 2.12).

ewencp reviewed (approved):
LGTM and verified that the core system tests still pass, including from 0.8.2 versions. Thanks @ppatierno!
